index.bs

```diff
@@ -2712,7 +2591,181 @@ partial interface MLGraphBuilder {
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: the type of |input| and |indices| is {{MLOperand}}.
+    1. Let |shapeInput| be |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}} and |rankInput| the [=list/size=] of |shapeInput|.
+    1. Let |shapeIndices| be |indices|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}}.
+    1. Let |axis| be |options|.{{MLGatherOptions/axis}}.
+    1. Let |axisSize| be |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}}[|axis|]
+    1. If |axis| is greater than or equal to |rankInput|, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. [=map/For each=] |index| &rarr; |value| of |indices|:
+        1. If |index| is greater than or equal to |axisSize|, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. Let |dimCount| be zero.
+    1. Let |rankOutput| be zero.
+    1. Let |shapeOutput| be an empty list.
+    1. [=map/For each=] |size| &rarr; |value| of |shapeInput|:
+        1. If |dimCount| is equal to |axis| then [=break=].
+        1. Set |shapeOutput|[|dimCount|] to |size|.
+        1. Increment |dimCount| by one.
+    1. Set |rankOutput| to |dimCount|.
+    1. Let |dimCount| be zero.
+    1. [=map/For each=] |size| &rarr; |value| of |shapeIndices|:
+        1. Set |shapeOutput|[|rankOutput| + |dimCount|] to |size|.
+        1. Increment |dimCount| by one.
+    1. Set |rankOutput| to |rankOutput| + |dimCount|.
+    1. Let |dimCount| be zero.
+    1. [=map/For each=] |size| &rarr; |value| of |shapeInput|:
+        1. If |dimCount| is less than or equal to |axis| then [=continue=].
+        1. Set |shapeOutput|[|rankOutput| + |dimCount|] to |size|.
+        1. Increment |dimCount| by one.
+    1. Let |desc| a new {{MLOperandDescriptor}}.
+    1. Set |desc|.{{MLOperandDescriptor/dimensions}} to |shapeOutput|.
+    1. Set |desc|.{{MLOperandDescriptor/type}} to |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/type}}.
+    1. If any of the following sub-steps fail, [=exception/throw=] an "{{OperationError}}" {{DOMException}}.
+        1. Let |output| be the result of <a>creating an MLOperand</a> given |desc|.
+        1. Make a request to the underlying platform to:
+            1. Let |opImpl| be an [=implementation-defined=] platform operator for the Gather operation, given |input|, |indices|, and |options|.
+            1. Store a reference of |opImpl| in |output|.{{MLOperand/[[operator]]}}.
+            1. Create an [=implementation-defined=] platform operand |outputImpl| to represent the output, given |output| and |opImpl|.
+            1. Store a reference to |outputImpl| in |output|.{{MLOperand/[[operand]]}}.
+        1. Connect |input|.{{MLOperand/[[operand]]}} and |indices|.{{MLOperand/[[operand]]}} as inputs to |opImpl|.
+        1. Connect |output|.{{MLOperand/[[operand]]}} as output to |opImpl|.
+    1. Return |output|.
+  </div>
+</details>
+
```

Comment:
Can we add a "implementation consideration" (i.e. an informative section) mentioning how out-of-bound indices should be handled by browser vendor and platform implementation?

Some bullet points:

1. Runtime out-of-bounds indices should be explicitly handled by either the browser implementation or the platform implementation, to avoid OOB memory accesses.
2. If the platform implementation doesn't handle out-of-bounds indices, the browser implementation should take steps to ensure the platform operator doesn't receive out-of-bound indices
3. Mention what caller should expect as a result of list item 2: 0, NaN, first/last indices (if implemented with clamp)

Behavior of 3 should ideally be standardized in a saperate issue. 
I'd expect some amount of discussion the best approach isn't obvious (see https://github.com/webmachinelearning/webnn/issues/375#issuecomment-1759263083)

Reply:
Opened a separate issue to track: https://github.com/webmachinelearning/webnn/issues/486